assets/app/view/game/part/track.rb

```diff
@@ -50,21 +88,47 @@ def render
             .flat_map { |path| path_indexes[path].map { |i| [path, i] } }
             .sort_by { |_, index| index || -1 }
 
-          sorted.map do |path, index|
+          # Main track
+          #
+          # Non-stub/offboard track requires up to three passes:
+          # - Draw borders
+          # - Draw the actual track
+          # - Draw the stripped inner bits (narrow gauge only)
+          #
+          # Each pass has to be grouped together so that track "switches" are rendered correctly
+          #
+          # Stubs and offboards can be done in a single pass because they don't have switches
+          #
+          pass0 = []
+          pass1 = []
+          pass2 = []
+          sorted.each do |path, index|
             props = {
-              color: value_for_index(index, :color, path.track),
+              color: value_for_index(index, :color, path.gauge),
               width: width_for_index(path, index, path_indexes),
-              dash: value_for_index(index, :dash, path.track),
+              dash: value_for_index(index, :dash, path.gauge),
             }
 
+            border_props = BORDER_PROPS[path.gauge]
+            inner_props = INNER_PROPS[path.gauge]
+
             if path.stub?
-              h(TrackStub, stub: path, region_use: @region_use, **props)
+              pass1 << h(TrackStub, stub: path, region_use: @region_use, border_props: border_props, **props)
             elsif path.offboard
-              h(TrackOffboard, offboard: path.offboard, path: path, region_use: @region_use, **props)
+              pass1 << h(TrackOffboard, offboard: path.offboard, path: path, region_use: @region_use,
+                                        border_props: border_props, **props)
             else
-              h(TrackNodePath, tile: @tile, path: path, region_use: @region_use, **props)
+              pass0 << h(TrackNodePath, tile: @tile, path: path, region_use: @region_use,
+                                        pass: 0, border_props: border_props, inner_props: inner_props, **props)
+              pass1 << h(TrackNodePath, tile: @tile, path: path, region_use: @region_use,
+                                        pass: 1, border_props: border_props, inner_props: inner_props, **props)
+              if inner_props
+                pass2 << h(TrackNodePath, tile: @tile, path: path, region_use: @region_use,
+                                          pass: 2, border_props: border_props, inner_props: inner_props, **props)
+              end
             end
           end
+          (pass0 + pass1 + pass2).compact
```

Comment:
pass0.concat(pass1).concat(pass2).compact! does it without allocation i think

Reply:
but why would we be adding null passes? can we prevent the nulls from getting added?

Reply:
Even better, thanks.

Reply:
Just need pass0.concat(pass1).concat(pass2)